### PR TITLE
Demonstrate how to create dynamic strategies with complex keys

### PR DESCRIPTION
**Pull request by martintmk**
### Details on the issue fix or feature implementation

This sample demonstrates how to define a HTTP resilience strategy that is created and configured for different combinations of endpoints/resources. 

The idea is to cache the strategy by complex `EndpointKey`. This sample also demonstrates how `EndpointKey` can be used to create dynamic strategies where some inner strategies are conditional.

Paired with the `EndpointKey` is the configuration of `ResilienceStrategyRegistryOptions<EndpointKey>` where we customize how the registry creates and caches the strategies.

@martincostello 

Contributes to #1365 

### Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

**Comment by martincostello** [comment]
Thanks for this - I'll have a play around with it now.

In my case, these options objects are bound from `IConfiguration` via DI, so they observe configuration changes. In this way of doing things, how would they observe changes? With the code I already have, we explicitly just clear the policy registry when the configuration changes and then policies are then rebuilt from scratch to account for any configuration change.

**Comment by martincostello** [comment]
It's also getting a bit bootstrap-y, as I need the options from configuration to configure the service container, as that's where the endpoint options come from.

**Comment by martincostello** [comment]
This is starting to get a bit complicated - tomorrow I'll try and extract the relevant parts of the v7 Polly code into a runnable-sample with the existing tests - then it'll be easier to see what the migration path would be and if there's anything we can do to make that easier.

**Comment by martintmk** [comment]
@martincostello 

> In my case, these options objects are bound from IConfiguration via DI, so they observe configuration changes. In this way of doing things, how would they observe changes?

In case you configure the `EndpointOptions` in DI the sample can be simplified. Instead of including it in the key you can just resolve those from DI and use it to configure resilience strategy. Additionally,  you can instruct Polly to dynamically reload the resilience strategies when these options are changed. 

Instead of:
``` csharp
var endpointOptions = context.StrategyKey.EndpointOptions;
```

You do:

``` csharp
// resolve the options from DI
var endpointOptions = context.GetOptions<EndpointOptions>(context.StrategyKey.EndpointName);

// enable dynamic reload of the resilience strategy being configured whenever the named options are changed
context.EnableReloads<EndpointOptions>(context.StrategyKey.EndpointName);
```

File: test/Polly.Extensions.Tests/Issues/IssuesTests.StrategiesPerEndpoint_1365.cs [hunk 1]
```diff
@@ -97,16 +97,25 @@ public void StrategiesPerEndpoint_1365()
...
         public int MaxParallelization { get; set; } = 10;
     }
 
-    public record EndpointKey(string BuilderName, string EndpointName, string Resource, EndpointOptions EndpointOptions)
+    public class EndpointsOptions
+    {
+        public Dictionary<string, EndpointOptions> Endpoints { get; set; } = new()
+        {
+            {  "Endpoint 1", new EndpointOptions { Retries = 2 } },
+            {  "Endpoint 2", new EndpointOptions { Retries = 3 } },
```
Comment (martincostello) [inline test/Polly.Extensions.Tests/Issues/IssuesTests.StrategiesPerEndpoint_1365.cs:114]:
```suggestion
            ["Endpoint 1"] = new() { Retries = 2 },
            ["Endpoint 2"] = new() { Retries = 3 },
```

File: test/Polly.Extensions.Tests/Issues/IssuesTests.StrategiesPerEndpoint_1365.cs [hunk 2]
```diff
@@ -16,18 +18,24 @@ public void StrategiesPerEndpoint_1365()
...
 
         services.AddResilienceStrategy<string>();
+        services.AddOptions<EndpointsOptions>();
 
         // add resilience strategy, keyed by EndpointKey that only defines the builder name
-        services.AddResilienceStrategy(new EndpointKey("endpoint-pipeline", string.Empty, string.Empty, new()), (builder, context) =>
+        services.AddResilienceStrategy(new EndpointKey("endpoint-pipeline", string.Empty, string.Empty), (builder, context) =>
         {
-            var endpointOptions = context.StrategyKey.EndpointOptions;
+            // we want this pipeline to react to changes in options
```
Comment (martincostello) [inline test/Polly.Extensions.Tests/Issues/IssuesTests.StrategiesPerEndpoint_1365.cs:26]:
```suggestion
            // we want this pipeline to react to changes to the options
```

File: test/Polly.Extensions.Tests/Issues/IssuesTests.StrategiesPerEndpoint_1365.cs [hunk 3]
```diff
@@ -16,18 +18,24 @@ public void StrategiesPerEndpoint_1365()
...
 
+            var endpointOptions = context.GetOptions<EndpointsOptions>().Endpoints[context.StrategyKey.EndpointName];
             var registry = context.ServiceProvider.GetRequiredService<ResilienceStrategyRegistry<string>>();
 
             // we want to limit the number of concurrent requests per endpoint and not include the resource.
             // using a registry we can create and cache the shared resilience strategy
-            var rateLimiterStrategy = registry.GetOrAddStrategy($"rate-limiter/{context.StrategyKey.EndpointName}", b =>
+            var rateLimiterStrategy = registry.GetOrAddStrategy($"rate-limiter/{context.StrategyKey.EndpointName}", (b, c) =>
             {
+                // let's also enable reloads for rate limiter
```
Comment (martincostello) [inline test/Polly.Extensions.Tests/Issues/IssuesTests.StrategiesPerEndpoint_1365.cs:36]:
```suggestion
                // let's also enable reloads for the rate limiter
```

File: test/Polly.Extensions.Tests/Issues/IssuesTests.StrategiesPerEndpoint_1365.cs [hunk 4]
```diff
@@ -16,18 +18,24 @@ public void StrategiesPerEndpoint_1365()
...
+            // we want this pipeline to react to changes in options
+            context.EnableReloads<EndpointsOptions>();
 
+            var endpointOptions = context.GetOptions<EndpointsOptions>().Endpoints[context.StrategyKey.EndpointName];
             var registry = context.ServiceProvider.GetRequiredService<ResilienceStrategyRegistry<string>>();
 
             // we want to limit the number of concurrent requests per endpoint and not include the resource.
             // using a registry we can create and cache the shared resilience strategy
-            var rateLimiterStrategy = registry.GetOrAddStrategy($"rate-limiter/{context.StrategyKey.EndpointName}", b =>
+            var rateLimiterStrategy = registry.GetOrAddStrategy($"rate-limiter/{context.StrategyKey.EndpointName}", (b, c) =>
```
Comment (martincostello) [inline test/Polly.Extensions.Tests/Issues/IssuesTests.StrategiesPerEndpoint_1365.cs:34]: Maybe we should make these more explicit as `builder` and `context`?

**Review by martincostello — APPROVED** [review]

**Comment by martintmk** [comment]
> Sure - btw have you forgotten to re-open Slack since your vacation? 😄

Ooops, I check it maybe once, twice a day so the risk of me missing something is high :D

**Pull request merged**